modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -233,6 +248,11 @@ private void analyzeWrite(final CompilerSettings settings, final Definition defi
     private void analyzeRead() {
         final ALink last = links.get(links.size() - 1);
 
+        // If the load node is a DEF node, we adapt its after type to use _this_ expected output type:
+        if (last instanceof ADefLink && this.expected != null) {
+            last.after = this.expected;
+        }
+
```

Comment:
I tried the cast variant, it breaks because JVM stack gets corrupt because of duping in compound assignments. So I reverted it. I will leave my solution for now.